### PR TITLE
PWN-7176 - New swap. Correct slippage error detect

### PR DESCRIPTION
**Pull request by gslevinkov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7176

## Description of Work

Used:
https://docs.rs/solana-sdk/latest/solana_sdk/transaction/enum.TransactionError.html
https://github.com/jup-ag/jupiter-cpi/blob/main/idl.json

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/infrastructure/network/data/ServerException.kt [hunk 1]
```diff
@@ -8,7 +8,8 @@ class ServerException(
     val errorCode: ErrorCode,
     fullMessage: String,
     private val errorMessage: String?,
-    val jsonErrorBody: JsonObject? = null
+    val jsonErrorBody: JsonObject? = null,
+    val domainErrorType: RpcTransactionError? = null
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/data/ServerException.kt:12]: doesn't break work with this class
still exception, but with additional nullable trycatched field

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/infrastructure/network/data/InstructionErrorType.kt [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.google.gson.JsonElement
+
+private const val KEY_CUSTOM_INSTRUCTION_ERROR_ID = "Custom"
+
+sealed interface InstructionErrorType {
+    companion object {
+        // can be { "Custom" : 1111 } or just plain "SomeError"
+        fun from(instructionErrorType: JsonElement): InstructionErrorType {
+            val isCustom = instructionErrorType.isJsonPrimitive
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/data/InstructionErrorType.kt:11]: Why if it's primitive we cast it as JsonObject ? if I am not wrong, String is primitive too.

File: app/src/main/java/org/p2p/wallet/infrastructure/network/data/RpcTransactionError.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ * }
+ * ```
+ * @return valid typed RpcTransactionError or null if parsing is failed
+ */
+sealed interface RpcTransactionError {
+    companion object {
+        fun from(errJson: JsonObject): RpcTransactionError? {
+            // for example "InstructionError"
+            val transactionErrorType: String? = errJson.keySet().firstOrNull()
+            return when (transactionErrorType) {
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/data/RpcTransactionError.kt:32]: you can wrap it to some enum class

File: app/src/main/java/org/p2p/wallet/infrastructure/network/data/RpcTransactionError.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                "AccountInUse" -> AccountInUse
+                "InstructionError" -> parseInstructionError(errJson, transactionErrorType)
+                else -> null
+            }
+        }
+    }
+
+    object AccountInUse : RpcTransactionError
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/data/RpcTransactionError.kt:40]: can AccountInUse and InstructionError be inherited from RpcTransactionError sealed class ?

File: app/src/main/java/org/p2p/wallet/jupiter/interactor/JupiterSwapInteractor.kt [hunk 5]
```diff
@@ -47,9 +46,10 @@ class JupiterSwapInteractor(
         Timber.i("Swap tokens success: $firstTransactionSignature")
         JupiterSwapTokensResult.Success
     } catch (error: ServerException) {
-        val errorMessage = error.message.orEmpty()
-        if (LOW_SLIPPAGE_ERROR_CODE in errorMessage ||
-            LOW_SLIPPAGE_ERROR_MESSAGE in errorMessage
+        if (error.domainErrorType != null &&
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/jupiter/interactor/JupiterSwapInteractor.kt:49]: you can move error.domainErrorType to variable

**Review by NikolaiEmelianov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/infrastructure/network/data/InstructionErrorType.kt (hunk 2 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/infrastructure/network/data/InstructionErrorType.kt:11]: thanks for pointing this out, made a test to check this stuff

File: app/src/main/java/org/p2p/wallet/infrastructure/network/data/RpcTransactionError.kt (hunk 3 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/infrastructure/network/data/RpcTransactionError.kt:32]: no need for that, we know all the string values

File: app/src/main/java/org/p2p/wallet/infrastructure/network/data/RpcTransactionError.kt (hunk 4 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/infrastructure/network/data/RpcTransactionError.kt:40]: already inherited, or i haven't catched your question

File: app/src/main/java/org/p2p/wallet/jupiter/interactor/JupiterSwapInteractor.kt (hunk 5 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/jupiter/interactor/JupiterSwapInteractor.kt:49]: done

**Pull request merged**